File: tests/test_wizard_password_logic.py
"""Smoke test for wizard password / continue-button logic (mirrors app.js)."""


def db_needs_password(db):
    return db in ("mysql", "mariadb", "postgresql", "timescaledb")


def password_candidates_confirmed(rows, confirmed, values):
    if not rows:
        return True
    return all(
        confirmed.get(r["key"]) and (values.get(r["key"]) or r.get("value") or "").strip()
        for r in rows
    )


def get_migration_password(db, rows, confirmed, values):
    for r in rows:
        if r.get("used_for_migration") and confirmed.get(r["key"]):
            val = (values.get(r["key"]) or r.get("value") or "").strip()
            if val:
                return val
    order = (
        ["MYSQL_ROOT_PASSWORD", "MYSQL_PASSWORD", "DB_PASSWORD"]
        if db in ("mysql", "mariadb")
        else ["POSTGRES_PASSWORD", "DB_PASSWORD"]
    )
    for key in order:
        if confirmed.get(key) and (values.get(key) or "").strip():
            return values[key].strip()
    return None


def has_db_credentials(db, rows, confirmed, values):
    if not db_needs_password(db):
        return True
    if not password_candidates_confirmed(rows, confirmed, values):
        return False
    return bool(get_migration_password(db, rows, confirmed, values))


def can_proceed_step2(source_db, rows, confirmed, values, upload_complete):
    if not source_db:
        return "no source db"
    if db_needs_password(source_db) and not has_db_credentials(source_db, rows, confirmed, values):
        if not password_candidates_confirmed(rows, confirmed, values):
            return "password not confirmed"
        return "creds incomplete"
    if not upload_complete:
        return "upload incomplete"
    return None


def test_wizard_password_flow():
    rows = [
        {"key": "MYSQL_ROOT_PASSWORD", "value": "rootpass", "used_for_migration": True},
        {"key": "DB_PASSWORD", "value": "dbpass", "used_for_migration": False},
    ]
    assert not password_candidates_confirmed(rows, {}, {})
    assert not password_candidates_confirmed(
        rows, {"MYSQL_ROOT_PASSWORD": True}, {"MYSQL_ROOT_PASSWORD": "rootpass"}
    )
    all_confirmed = {"MYSQL_ROOT_PASSWORD": True, "DB_PASSWORD": True}
    all_values = {"MYSQL_ROOT_PASSWORD": "rootpass", "DB_PASSWORD": "dbpass"}
    assert password_candidates_confirmed(rows, all_confirmed, all_values)
    assert has_db_credentials("mysql", rows, all_confirmed, all_values)
    assert can_proceed_step2("mysql", rows, all_confirmed, all_values, True) is None
    assert can_proceed_step2(
        "mysql",
        rows,
        {"MYSQL_ROOT_PASSWORD": True},
        {"MYSQL_ROOT_PASSWORD": "rootpass"},
        True,
    ) is not None
    print("OK: wizard password flow")


if __name__ == "__main__":
    test_wizard_password_flow()
    print("All wizard password logic tests passed")

File: tests/test_wizard_password_logic.js
/**
 * Node smoke test for wizard password / continue-button logic (mirrors app.js).
 * Run: node tests/test_wizard_password_logic.js
 */

function dbNeedsPassword(db) {
  return db === 'mysql' || db === 'mariadb' || db === 'postgresql' || db === 'timescaledb';
}

function passwordCandidatesConfirmed(rows, confirmed, values) {
  if (!rows.length) return true;
  return rows.every(r => confirmed[r.key] && (values[r.key] ?? r.value ?? '').trim());
}

function getMigrationPassword(db, rows, confirmed, values) {
  const migrationRow = rows.find(r => r.used_for_migration && confirmed[r.key]);
  if (migrationRow) {
    const val = (values[migrationRow.key] ?? migrationRow.value ?? '').trim();
    if (val) return val;
  }
  const order = db === 'mysql' || db === 'mariadb'
    ? ['MYSQL_ROOT_PASSWORD', 'MYSQL_PASSWORD', 'DB_PASSWORD']
    : ['POSTGRES_PASSWORD', 'DB_PASSWORD'];
  for (const key of order) {
    if (confirmed[key] && (values[key] ?? '').trim()) return values[key].trim();
  }
  return null;
}

function hasDbCredentials(db, rows, confirmed, values) {
  if (!dbNeedsPassword(db)) return true;
  if (!passwordCandidatesConfirmed(rows, confirmed, values)) return false;
  return !!getMigrationPassword(db, rows, confirmed, values);
}

function canProceedStep2({ sourceDb, rows, confirmed, values, uploadComplete }) {
  if (!sourceDb) return 'no source db';
  if (dbNeedsPassword(sourceDb) && !hasDbCredentials(sourceDb, rows, confirmed, values)) {
    if (!passwordCandidatesConfirmed(rows, confirmed, values)) return 'password not confirmed';
    return 'creds incomplete';
  }
  if (!uploadComplete) return 'upload incomplete';
  return null;
}

let failed = 0;
function assert(cond, msg) {
  if (!cond) {
    console.error('FAIL:', msg);
    failed++;
  } else {
    console.log('OK:', msg);
  }
}

const rows = [
  { key: 'MYSQL_ROOT_PASSWORD', value: 'rootpass', used_for_migration: true },
  { key: 'DB_PASSWORD', value: 'dbpass', used_for_migration: false },
];

// Before fix this threw ReferenceError: passwordCandidatesConfirmed is not defined
assert(typeof passwordCandidatesConfirmed === 'function', 'passwordCandidatesConfirmed exists');

const emptyConfirmed = {};
const emptyValues = {};
assert(!passwordCandidatesConfirmed(rows, emptyConfirmed, emptyValues), 'unconfirmed passwords block');

const oneConfirmed = { MYSQL_ROOT_PASSWORD: true };
const oneValues = { MYSQL_ROOT_PASSWORD: 'rootpass' };
assert(!passwordCandidatesConfirmed(rows, oneConfirmed, oneValues), 'partial confirm blocks');

const allConfirmed = { MYSQL_ROOT_PASSWORD: true, DB_PASSWORD: true };
const allValues = { MYSQL_ROOT_PASSWORD: 'rootpass', DB_PASSWORD: 'dbpass' };
assert(passwordCandidatesConfirmed(rows, allConfirmed, allValues), 'all passwords confirmed');

assert(
  hasDbCredentials('mysql', rows, allConfirmed, allValues),
  'hasDbCredentials true when migration password confirmed'
);

assert(
  canProceedStep2({
    sourceDb: 'mysql',
    rows,
    confirmed: allConfirmed,
    values: allValues,
    uploadComplete: true,
  }) === null,
  'can proceed step2 after confirm + upload'
);

assert(
  canProceedStep2({
    sourceDb: 'mysql',
    rows,
    confirmed: oneConfirmed,
    values: oneValues,
    uploadComplete: true,
  }) !== null,
  'cannot proceed with only one password confirmed'
);

if (failed) {
  console.error(`\n${failed} test(s) failed`);
  process.exit(1);
}
console.log('\nAll wizard password logic tests passed');
